UpdateUserGroups/handler.ts

```diff
@@ -0,0 +1,381 @@
+  IResponseSuccessJson,
+  ResponseErrorNotFound,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { EmailAddress } from "../generated/definitions/EmailAddress";
+import { GroupCollection } from "../generated/definitions/GroupCollection";
+import { UserGroupsPayload } from "../generated/definitions/UserGroupsPayload";
+import {
+  getApiClient,
+  IAzureApimConfig,
+  IServicePrincipalCreds
+} from "../utils/apim";
+import { groupContractToApiGroup } from "../utils/conversions";
+import { genericInternalErrorHandler } from "../utils/errorHandler";
+
+type IGetSubscriptionKeysHandlerResponseError =
+  | IResponseErrorNotFound
+  | IResponseErrorInternal
+  | IResponseErrorValidation;
+type IGetSubscriptionKeysHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  clientIp: ClientIp,
+  userAttributes: IAzureUserAttributes,
+  email: EmailAddress,
+  userGroupsPayload: UserGroupsPayload
+) => Promise<
+  | IResponseSuccessJson<GroupCollection>
+  | IGetSubscriptionKeysHandlerResponseError
+>;
+
+function getGroups(
+  apimClient: ApiManagementClient,
+  apimResourceGroup: string,
+  apim: string
+): TaskEither<Error, ReadonlyArray<GroupContract>> {
+  return tryCatch(async () => {
+    // tslint:disable-next-line:readonly-array no-let
+    const groupList: GroupContract[] = [];
+    const groupListResponse = await apimClient.group.listByService(
+      apimResourceGroup,
+      apim
+    );
+    groupList.push(...groupListResponse);
+    // tslint:disable-next-line:no-let
+    let nextLink = groupListResponse.nextLink;
+    while (nextLink) {
+      const nextGroupList = await apimClient.group.listByServiceNext(nextLink);
+      groupList.push(...nextGroupList);
+      nextLink = nextGroupList.nextLink;
+    }
+    return groupList;
+  }, toError);
+}
+
+function getUserGroups(
+  apimClient: ApiManagementClient,
+  apimResourceGroup: string,
+  apim: string,
+  userName: string
+): TaskEither<Error, ReadonlyArray<GroupContract>> {
+  return tryCatch(async () => {
+    // tslint:disable-next-line:readonly-array no-let
+    const groupList: GroupContract[] = [];
+    const groupListResponse = await apimClient.userGroup.list(
+      apimResourceGroup,
+      apim,
+      userName
+    );
+    groupList.push(...groupListResponse);
+    // tslint:disable-next-line:no-let
+    let nextLink = groupListResponse.nextLink;
+    while (nextLink) {
+      const nextGroupList = await apimClient.userGroup.listNext(nextLink);
+      groupList.push(...nextGroupList);
+      nextLink = nextGroupList.nextLink;
+    }
+    return groupList;
+  }, toError);
+}
+
+interface IGroupsClusterization {
+  toBeAssociated: ReadonlyArray<string>;
+  toBeRemoved: ReadonlyArray<string>;
+}
+
+function clusterizeGroups(
+  existingGroups: Record<string, string>,
+  currentUserGroups: ReadonlyArray<string>,
+  groupsInPayload: ReadonlyArray<string>
+): IGroupsClusterization {
+  return new StrMap(existingGroups).reduceWithKey(
```

Comment:
Maybe the `partion` method of StrMap could help here instead of reduce (or partitionMap)

Reply:
I'm not sure about this... In the current implementation, only some of the existing groups are added to the clusterization, that is the groups with which the user must be associated and the groups from which the user must be deleted: if the user was not associated with a group and still must not be associated, then the group isn't included into the clusterization; the same for a group with which the user is already associated and with which must still be associated. If I use the partition `method`, instead, all the groups will be added to the clusterization, so an operation from the apim client will be performed for each of them even if it this is not necessary.